src/internal/Internal/SafeNatural.hs

```diff
@@ -0,0 +1,130 @@
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DerivingVia #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+
+module Internal.SafeNatural
```

Comment:
Isn't the point of `Coin` to be the `SafeNatural` :grimacing: ? We may be over-doing it here.

Reply:
@KtorZ wrote:
> Isn't the point of `Coin` to be the `SafeNatural` grimacing ?

That's a good question. `SafeNatural` is actually used as the underlying representation of two types:

- `Coin`
- `DustThreshold`

This is partly because it doesn't make sense for a dust threshold to be negative, and partly for convenience, as much of the time we are comparing dust thresholds with coin values, and it helps to avoid unnecessary conversions.

Since the `SafeNatural` type is encapsulated in its own module, and since that module only exports smart constructors and total functions on `SafeNatural`, it should be relatively hard to accidentally create negative values of `Coin`, due to the protection conferred by the underlying type.

Reply:
`DustThreshold` used to be expressed in `Coin`, so by definition couldn't possibly be negative. I still fail to see what value do we gain by having this extra indirection to `SafeNatural` which requires redefining all the internals of an already quite idiomatic type of the Haskell eco-system: `Natural` :( .

Reply:
@KtorZ wrote:
> I still fail to see what value do we gain by having this extra indirection to SafeNatural which requires redefining all the internals of an already quite idiomatic type of the Haskell eco-system: Natural

The issue with Natural is that it provides a `Num` instance, which requires an implementation of the `-` operator:
```hs
(-) :: a -> a -> a
```

Just having this operator available means that it's trivial for a programmer (several months down the line) to accidentally introduce an **_unchecked_** subtraction back into the code.

Of course, good code review and good test coverage should ideally stop this from happening, but sometimes it doesn't. (See https://github.com/input-output-hk/cardano-coin-selection/pull/50#discussion_r411138543 for a great example.)

Reply:
I get your point but I still disagree. I think it's pushing the complexity line a bit too far for only little added benefits. This isn't code that is changing _much_ and, there are all sort of manipulation with the fee and coins to be careful with already. So it's not like we take substraction and its consequences out of consideration in this code. 

I am quite curious though about the added complexity pushed onto the library users. If the added complexity is only visible to us internally then.. whatever. But if this results in an extra cost for library users, then it's a no go.

Reply:
@KtorZ wrote:

> I am quite curious though about the added complexity pushed onto the library users. If the added complexity is only visible to us internally then.. whatever. But if this results in an extra cost for library users, then it's a no go.

I don't think that a user will see much complexity, as it's deliberately hidden from view.

For converting a coin to some other type (after running a coin selection algorithm) we provide:
```hs
coinToIntegral :: Integral i -> Coin -> i
```
For constructing values of type `Coin` (in order to pass in to a coin selection algorithm), we provide a smart constructor:
```hs
coin :: Integral i -> Maybe Coin
```
We **_could_** make this even easier by giving a choice of functions:
```hs
coinFromIntegral :: Integral i -> Maybe Coin
coinFromNatural  :: Natural    ->       Coin
```
Thoughts?

Reply:
why not to encapsulate everything in `Coin` and then use it inside `Fee` and `DustThreshold` ? (ie., resign from `SafeNatural` in a sense that `Coin` would have it, and then add `Num` instance for `Coin` :thinking:

Reply:
@paweljakubas wrote:
> then add Num instance for Coin

If we provide a `Num` instance for `Coin`, then it opens the door to creating coins with negative values. This is because the `Num` class requires an implementation of `-`, which has the following type:

```hs
(-) :: a -> a -> a
```

The only way to provide a definition that satisfies this type without allowing negative coin values is to throw a run-time exception if the second argument is greater than the first.

But the `SafeNatural` type makes this **impossible** by _not_ providing a `Num` instance, instead providing:

```hs
sub      :: SafeNatural -> SafeNatural -> Maybe SafeNatural
distance :: SafeNatural -> SafeNatural ->       SafeNatural
```

Of course, we _could_ merge the definition of `Coin` and `SafeNatural`, and then redefine `Fee` and `DustThreshold` in terms of `Coin`, but this really wouldn't change much from the user's point of view, if at all.

Reply:
and in case of impl `(-)` of `Num` we cannot have `assert` to outlaw negative values? We have instances of `Num` for `GHC.Natural` implemented, like here https://hackage.haskell.org/package/base-4.12.0.0/docs/src/GHC.Num.html#line-143 We can maybe have in case of `(-)` some customized mechanism or fallback on what `Natural` offers? No?

Reply:
@paweljakubas wrote:
> why not to encapsulate everything in Coin and then use it inside Fee and DustThreshold 

Fixed in 01a3a62160221f5ab8adf98d6e11de1fac611147.

Reply:
@KtorZ The public API now exposes the following constructor (and deconstructor):
```hs
coinFromNatural :: Natural -> Coin

coinToNatural :: Coin -> Natural
```
Hopefully this should be fairly easy to use, but let me know if you'd prefer a different way.

(Internally, we also have `coinFromIntegral`, but this isn't exposed in the public API.)

Reply:
@KtorZ Additionally, the definition of `SafeNatural` is now merged into `Coin`.

`Coin` is now a `newtype` wrapper around `Natural`. We retain the set of safe operations, as before.

`Fee` and `DustThreshold` are now just `newtype` wrappers around `Coin`.